File: src/cf_ip_blacklist/policy.py
from __future__ import annotations

import ipaddress
from datetime import datetime, timedelta

from .config import Settings
from .models import IPRecord, Observation


def normalize_ip(value: str) -> str:
    try:
        return str(ipaddress.ip_address(value.strip()))
    except ValueError as exc:
        raise ValueError(f"invalid IP address: {value}") from exc


def load_allowlist(path: str) -> tuple[ipaddress.IPv4Network | ipaddress.IPv6Network, ...]:
    networks: list[ipaddress.IPv4Network | ipaddress.IPv6Network] = []
    with open(path) as handle:
        for line_number, raw in enumerate(handle, start=1):
            value = raw.split("#", 1)[0].strip()
            if not value:
                continue
            try:
                networks.append(ipaddress.ip_network(value, strict=False))
            except ValueError as exc:
                raise ValueError(f"invalid allowlist entry at line {line_number}: {value}") from exc
    return tuple(networks)


def is_allowlisted(
    ip: str, networks: tuple[ipaddress.IPv4Network | ipaddress.IPv6Network, ...]
) -> bool:
    address = ipaddress.ip_address(ip)
    return any(address in network for network in networks)


def evaluate_observations(
    observations: list[Observation], existing: IPRecord | None, settings: Settings, now: datetime
) -> IPRecord:
    if not observations and existing is None:
        raise ValueError("cannot evaluate an empty observation set without existing state")
    first_seen = min(
        (o.observed_at for o in observations), default=existing.first_seen if existing else now
    )
    last_seen = max(
        (o.observed_at for o in observations), default=existing.last_seen if existing else now
    )
    observed = sum(o.observed_requests for o in observations) + (
        existing.observed_requests if existing else 0
    )
    weighted = sum(o.weighted_requests for o in observations) + (
        existing.weighted_requests if existing else 0
    )
    paths = {path for observation in observations for path in observation.paths}
    distinct_paths = max(len(paths), existing.distinct_paths if existing else 0)
    suspicious = sum(o.suspicious_paths for o in observations) + (
        existing.suspicious_paths if existing else 0
    )
    errors = sum(o.error_requests for o in observations) + (
        existing.error_requests if existing else 0
    )
    zones = sorted(
        {o.zone_id for o in observations} | set(existing.source_zones if existing else [])
    )
    windows = (existing.observation_windows if existing else 0) + (1 if observations else 0)
    error_ratio = errors / observed if observed else 0
    reasons: list[str] = []
    score = 0.0
    weights = settings.signal_weights
    if weighted >= settings.thresholds.min_weighted_requests:
        score += weights.request_volume
        reasons.append("request_volume")
    if distinct_paths >= settings.thresholds.min_distinct_paths:
        score += weights.path_breadth
        reasons.append("path_breadth")
    if suspicious >= settings.thresholds.min_suspicious_paths:
        score += weights.suspicious_paths
        reasons.append("suspicious_paths")
    if error_ratio >= settings.thresholds.max_error_ratio and observed > 0:
        score += weights.high_error_ratio
        reasons.append("high_error_ratio")
    if windows >= 2:
        score += weights.repeated_windows
        reasons.append("repeated_windows")
    if len(zones) >= 2:
        score += weights.multiple_zones
        reasons.append("multiple_zones")
    status = existing.status if existing else "candidate"
    blocked_at = existing.block_started_at if existing else None
    expires_at = existing.expires_at if existing else None
    extensions = existing.ttl_extensions if existing else 0
    if (
        score >= settings.thresholds.block_score
        and len(reasons) >= 2
        and suspicious >= settings.thresholds.min_suspicious_paths
    ):
        status = "blocked"
        blocked_at = blocked_at or now
        if expires_at is None:
            expires_at = now + timedelta(hours=settings.block_ttl_hours)
    elif status == "blocked":
        status = "candidate"
        blocked_at = None
        expires_at = None
    return IPRecord(
        first_seen=first_seen,
        last_seen=last_seen,
        last_evaluated=now,
        observed_requests=observed,
        weighted_requests=weighted,
        distinct_paths=distinct_paths,
        suspicious_paths=suspicious,
        error_requests=errors,
        observation_windows=windows,
        source_zones=zones,
        score=round(score, 4),
        status=status,
        reason_codes=sorted(set(reasons)),
        block_started_at=blocked_at,
        expires_at=expires_at,
        ttl_extensions=extensions,
    )

File: src/policy.rs
use crate::{
    config::{Settings, load_subject_file},
    error::{BlockholeError, Result},
    models::{IpRecord, Observation, RecordStatus, Subject},
};
use chrono::{DateTime, Utc};
use std::{collections::BTreeSet, path::Path};
pub fn allowlist(root: &Path) -> Result<Vec<Subject>> {
    load_subject_file(&root.join("config/allowlist.txt"))
}
pub fn permanent(root: &Path) -> Result<Vec<Subject>> {
    load_subject_file(&root.join("config/permanent-blocklist.txt"))
}
pub fn is_allowlisted(subject: &Subject, list: &[Subject]) -> bool {
    list.iter().any(|network| network.contains(subject))
}
/// Aggregated observation counters and signal scores, without status or decay.
pub struct MergedSignals {
    pub first_seen: DateTime<Utc>,
    pub last_seen: DateTime<Utc>,
    pub observed_requests: u64,
    pub weighted_requests: f64,
    pub distinct_paths: u64,
    pub suspicious_paths: u64,
    pub error_requests: u64,
    pub observation_windows: u64,
    pub source_zones: Vec<String>,
    pub raw_score: f64,
    pub reason_codes: Vec<String>,
    pub qualifies_for_block: bool,
}

/// Merge observations with existing record counters and compute signal scores.
///
/// Returns an error if both `observations` and `existing` are empty/None.
/// Does not determine status, apply decay, or set `last_evaluated`.
pub fn score_signals(
    observations: &[Observation],
    existing: Option<&IpRecord>,
    settings: &Settings,
    now: DateTime<Utc>,
) -> Result<MergedSignals> {
    if observations.is_empty() && existing.is_none() {
        return Err(BlockholeError::Policy(
            "cannot evaluate empty observations without state".into(),
        ));
    }
    let first_seen = observations
        .iter()
        .map(|o| o.observed_at)
        .min()
        .or_else(|| existing.map(|r| r.first_seen))
        .unwrap_or(now);
    let last_seen = observations
        .iter()
        .map(|o| o.observed_at)
        .max()
        .or_else(|| existing.map(|r| r.last_seen))
        .unwrap_or(now);
    let observed = observations
        .iter()
        .map(|o| o.observed_requests)
        .sum::<u64>()
        + existing.map_or(0, |r| r.observed_requests);
    let weighted = observations
        .iter()
        .map(|o| o.weighted_requests)
        .sum::<f64>()
        + existing.map_or(0.0, |r| r.weighted_requests);
    let paths: BTreeSet<_> = observations
        .iter()
        .flat_map(|o| o.paths.iter().cloned())
        .collect();
    let distinct = paths.len() as u64;
    let distinct = distinct.max(existing.map_or(0, |r| r.distinct_paths));
    let suspicious = observations.iter().map(|o| o.suspicious_paths).sum::<u64>()
        + existing.map_or(0, |r| r.suspicious_paths);
    let errors = observations.iter().map(|o| o.error_requests).sum::<u64>()
        + existing.map_or(0, |r| r.error_requests);
    let mut zones: BTreeSet<String> = observations.iter().map(|o| o.zone_id.clone()).collect();
    zones.extend(existing.map_or_else(Vec::new, |r| r.source_zones.clone()));
    let windows =
        existing.map_or(0, |r| r.observation_windows) + u64::from(!observations.is_empty());
    let ratio = if observed == 0 {
        0.0
    } else {
        errors as f64 / observed as f64
    };
    let mut reasons = Vec::new();
    let mut score = 0.0;
    let w = &settings.weights;
    if weighted >= settings.thresholds.min_weighted_requests {
        score += w.request_volume;
        reasons.push("request_volume".into());
    }
    if distinct >= settings.thresholds.min_distinct_paths {
        score += w.path_breadth;
        reasons.push("path_breadth".into());
    }
    if suspicious >= settings.thresholds.min_suspicious_paths {
        score += w.suspicious_paths;
        reasons.push("suspicious_paths".into());
    }
    if ratio >= settings.thresholds.max_error_ratio && observed > 0 {
        score += w.high_error_ratio;
        reasons.push("high_error_ratio".into());
    }
    if windows >= 2 {
        score += w.repeated_windows;
        reasons.push("repeated_windows".into());
    }
    if zones.len() >= 2 {
        score += w.multiple_zones;
        reasons.push("multiple_zones".into());
    }
    let qualifies = score >= settings.thresholds.block_score
        && reasons.len() >= 2
        && suspicious >= settings.thresholds.min_suspicious_paths;
    reasons.sort();
    reasons.dedup();
    Ok(MergedSignals {
        first_seen,
        last_seen,
        observed_requests: observed,
        weighted_requests: weighted,
        distinct_paths: distinct,
        suspicious_paths: suspicious,
        error_requests: errors,
        observation_windows: windows,
        source_zones: zones.into_iter().collect(),
        raw_score: (score * 10_000.0).round() / 10_000.0,
        reason_codes: reasons,
        qualifies_for_block: qualifies,
    })
}
pub fn merge_permanent(state: &mut crate::models::State, subjects: &[Subject], now: DateTime<Utc>) {
    let wanted: BTreeSet<_> = subjects.iter().cloned().collect();
    state.records.retain(|subject, record| {
        !matches!(record.status, RecordStatus::PermanentBlocked { .. }) || wanted.contains(subject)
    });
    for subject in subjects {
        let existing_suppressed = state
            .records
            .get(subject)
            .and_then(|r| match &r.status {
                RecordStatus::PermanentBlocked {
                    suppressed_by_allowlist,
                    ..
                } => Some(*suppressed_by_allowlist),
                _ => None,
            })
            .unwrap_or(false);

        let entry = state
            .records
            .entry(subject.clone())
            .or_insert_with(|| IpRecord {
                schema_version: crate::state::CURRENT_SCHEMA,
                first_seen: now,
                last_seen: now,
                last_evaluated: now,
                observed_requests: 0,
                weighted_requests: 0.0,
                distinct_paths: 0,
                suspicious_paths: 0,
                error_requests: 0,
                observation_windows: 0,
                source_zones: Vec::new(),
                score: 0.0,
                reason_codes: vec!["manual_import".into()],
                status: RecordStatus::PermanentBlocked {
                    imported_at: now,
                    source: "config/permanent-blocklist.txt".into(),
                    reason: None,
                    suppressed_by_allowlist: existing_suppressed,
                },
            });
        entry.status = RecordStatus::PermanentBlocked {
            imported_at: now,
            source: "config/permanent-blocklist.txt".into(),
            reason: None,
            suppressed_by_allowlist: existing_suppressed,
        };
    }
}
